File: uis/client/web2/src/services/wealth_service.ts
import { apiService } from './api_service';

const BASE = '/wealth/v1';

export interface WealthPortfolio {
  id: string;
  client?: string;
  client_name?: string;
  aum?: number | string;
  strategy?: string;
  ytd_return?: number | string;
  ytdReturn?: string;
  risk_profile?: string;
  riskProfile?: string;
  status?: string;
  [key: string]: unknown;
}

export interface PortfolioHolding {
  id: string;
  asset_name: string;
  ticker?: string;
  quantity: number;
  current_value: number | string;
  weight_pct?: number;
  [key: string]: unknown;
}

export interface PortfolioPerformance {
  portfolio_id: string;
  ytd_return?: number;
  one_year_return?: number;
  three_year_return?: number;
  since_inception?: number;
  [key: string]: unknown;
}

export interface WealthClient {
  id: string;
  name: string;
  tier?: string;
  relationship_manager?: string;
  [key: string]: unknown;
}

export interface InvestmentOption {
  id: string;
  name: string;
  strategy: string;
  min_investment?: number | string;
  expected_return?: string;
  risk_level?: string;
  [key: string]: unknown;
}

export interface WealthDashboard {
  total_aum?: number | string;
  total_portfolios?: number;
  active_clients?: number;
  avg_ytd_return?: number | string;
  [key: string]: unknown;
}

interface ApiResult<T> {
  success: boolean;
  data: T | null;
  message: string;
}

async function call<T>(fn: () => Promise<{ data: unknown }>): Promise<ApiResult<T>> {
  try {
    const res = await fn();
    const payload = res.data as { data?: T; results?: T } | T;
    const data =
      payload && typeof payload === 'object' && ('data' in payload || 'results' in payload)
        ? ((payload as { data?: T; results?: T }).data ?? (payload as { results?: T }).results ?? (payload as T))
        : (payload as T);
    return { success: true, data, message: 'OK' };
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : 'Request failed';
    return { success: false, data: null, message: msg };
  }
}

class WealthService {
  getPortfolios() {
    return call<WealthPortfolio[]>(() => apiService.get(`${BASE}/portfolios`));
  }

  getPortfolioById(id: string) {
    return call<WealthPortfolio>(() => apiService.get(`${BASE}/portfolios/${id}`));
  }

  createPortfolio(data: Record<string, unknown>) {
    return call<WealthPortfolio>(() => apiService.post(`${BASE}/portfolios`, data));
  }

  updatePortfolio(id: string, data: Record<string, unknown>) {
    return call<WealthPortfolio>(() => apiService.put(`${BASE}/portfolios/${id}`, data));
  }

  getPortfolioHoldings(id: string) {
    return call<PortfolioHolding[]>(() => apiService.get(`${BASE}/portfolios/${id}/holdings`));
  }

  getPortfolioPerformance(id: string) {
    return call<PortfolioPerformance>(() => apiService.get(`${BASE}/portfolios/${id}/performance`));
  }

  getClients() {
    return call<WealthClient[]>(() => apiService.get(`${BASE}/clients`));
  }

  getInvestmentOptions() {
    return call<InvestmentOption[]>(() => apiService.get(`${BASE}/investment-options`));
  }

  getDashboard() {
    return call<WealthDashboard>(() => apiService.get(`${BASE}/dashboard`));
  }
}

export const wealthService = new WealthService();

File: uis/client/mobile_app/lib/services/wealth_service.dart
import '../models/wealth_portfolio.dart';
import '../config/app_config.dart';
import 'api_service.dart';

class WealthService {
  final ApiService _apiService;

  WealthService(this._apiService);

  static const String _base = AppConfig.wealthEndpoint;

  Future<List<WealthPortfolio>> getPortfolios({
    int page = 1,
    int limit = 20,
    String? status,
    String? strategy,
  }) async {
    final params = <String, dynamic>{'page': page, 'limit': limit};
    if (status != null) params['status'] = status;
    if (strategy != null) params['strategy'] = strategy;

    final response = await _apiService.get('$_base/v1/portfolios', queryParameters: params);
    final data = response.data;
    final rawItems = data is Map ? (data['items'] ?? data['data'] ?? []) : data;
    if (rawItems is! List) return [];
    return rawItems
        .whereType<Map<String, dynamic>>()
        .map(WealthPortfolio.fromJson)
        .toList();
  }

  Future<WealthPortfolio> getPortfolioById(String id) async {
    final response = await _apiService.get('$_base/v1/portfolios/$id');
    return WealthPortfolio.fromJson(Map<String, dynamic>.from(response.data));
  }

  Future<WealthPortfolio> createPortfolio(Map<String, dynamic> body) async {
    final response = await _apiService.post('$_base/v1/portfolios', data: body);
    return WealthPortfolio.fromJson(Map<String, dynamic>.from(response.data));
  }

  Future<WealthPortfolio> updatePortfolio(String id, Map<String, dynamic> body) async {
    final response = await _apiService.put('$_base/v1/portfolios/$id', data: body);
    return WealthPortfolio.fromJson(Map<String, dynamic>.from(response.data));
  }

  Future<List<Map<String, dynamic>>> getPortfolioHoldings(String id) async {
    final response = await _apiService.get('$_base/v1/portfolios/$id/holdings');
    final data = response.data;
    final rawItems = data is Map ? (data['items'] ?? data['data'] ?? []) : data;
    if (rawItems is! List) return [];
    return rawItems.whereType<Map<String, dynamic>>().toList();
  }

  Future<List<Map<String, dynamic>>> getPortfolioPerformance(String id, {String? from, String? to}) async {
    final params = <String, dynamic>{};
    if (from != null) params['from'] = from;
    if (to != null) params['to'] = to;
    final response = await _apiService.get('$_base/v1/portfolios/$id/performance', queryParameters: params);
    final data = response.data;
    final rawItems = data is Map ? (data['items'] ?? data['data'] ?? []) : data;
    if (rawItems is! List) return [];
    return rawItems.whereType<Map<String, dynamic>>().toList();
  }

  Future<List<Map<String, dynamic>>> getClients({int page = 1, int limit = 20}) async {
    final response = await _apiService.get('$_base/v1/clients', queryParameters: {
      'page': page,
      'limit': limit,
    });
    final data = response.data;
    final rawItems = data is Map ? (data['items'] ?? data['data'] ?? []) : data;
    if (rawItems is! List) return [];
    return rawItems.whereType<Map<String, dynamic>>().toList();
  }

  Future<Map<String, dynamic>> getInvestmentOptions() async {
    final response = await _apiService.get('$_base/v1/investment-options');
    return Map<String, dynamic>.from(response.data);
  }

  Future<Map<String, dynamic>> getDashboard() async {
    final response = await _apiService.get('$_base/v1/dashboard');
    return Map<String, dynamic>.from(response.data);
  }
}
